File: lua/distract/render.lua
--- What the renderer draws, and how.
---
--- Mirrors `engine/src/render.rs`: one settings block, validated once here and
--- pushed to the overlay, so the terminal backends and the overlay read the same
--- numbers rather than each deciding for itself. That is the rule every measured
--- quantity in this plugin already follows — the floor, the viewport scope and the
--- obstacle list are all measured in Neovim and pushed to both engines.

local M = {}

local voxel = require("distract.voxel")

--- Draw one textured quad per entity, ordered by `z_index`. What every asset has
--- always been drawn as, and still the default.
M.FLAT = "2d"
--- Draw a voxel model extruded from the entity's current frame, depth-tested.
M.VOXEL = "3d"

local MODES = { [M.FLAT] = true, [M.VOXEL] = true }

---@class DistractLightConfig
---@field direction? number[] Direction the single directional light travels in [x, y, z] world axes (y is down)
---@field ambient? number Ambient floor brightness between 0.0 and 1.0 (default: 0.42)

---@class DistractRenderConfig
---@field mode? "2d"|"3d" Render mode: "2d" for flat sprites, "3d" for voxel models (default: "2d")
---@field fov_y_degrees? number Overlay only: vertical field of view in degrees (10 to 120, default: 45.0)
---@field depth_per_unit? number Depth of one unit of z as fraction of eye distance (0 to 0.5, default: 0.05)
---@field yaw_degrees? number Off-axis model rotation in degrees (default: 22.0)
---@field voxel_max_width? integer Widest voxel grid extruded before resampling (1 to 128, default: 48)
---@field voxel_depth? integer Slab thickness in voxels (1 to 64, default: 4)
---@field light? DistractLightConfig Directional and ambient lighting settings

---@type DistractRenderConfig
M.DEFAULTS = {
  mode = M.FLAT,
  fov_y_degrees = 45.0,
  --- Depth of one unit of `z`, as a fraction of the eye distance.
  depth_per_unit = 0.05,
  --- How far a model is turned off head-on, in degrees.
  ---
  --- Zero renders a model face-on, where it is indistinguishable from its own
  --- sprite: the depth is there and nothing reveals it.
  yaw_degrees = 22.0,
  voxel_max_width = voxel.DEFAULT_MAX_WIDTH,
  voxel_depth = voxel.DEFAULT_DEPTH,
  light = {
    --- Direction the light travels, in world axes. `y` is down, so a positive `y`
    --- is a light from above.
    direction = { -0.4, 0.8, -0.45 },
    --- Floor brightness for a face the light does not reach. A face in full
    --- shadow at 0 is pure black, which reads as a hole in the model.
    ambient = 0.42,
  },
}

--- Bounds every field is held to, because they come from a user's configuration.
M.MIN_FOV_Y_DEGREES = 10.0
M.MAX_FOV_Y_DEGREES = 120.0
M.MAX_VOXEL_MAX_WIDTH = 128
M.MAX_VOXEL_DEPTH = 64
M.MAX_DEPTH_PER_UNIT = 0.5

local function clamp(value, minimum, maximum)
  return math.max(minimum, math.min(maximum, value))
end

local function require_number(field, value)
  if type(value) ~= "number" then
    error(string.format("distract: render.%s must be a number", field))
  end
  return value
end

--- Validates and clamps a user's `render` block.
---
--- Refuses an unknown mode outright rather than falling back to a default: a
--- typo in `mode = "3D "` that silently rendered 2D is indistinguishable from a
--- renderer that does not work.
---@param config table|nil the user's `render` table
---@return table settings
function M.settings(config)
  local merged = vim.tbl_deep_extend("force", vim.deepcopy(M.DEFAULTS), config or {})

  if not MODES[merged.mode] then
    error(
      string.format(
        "distract: render.mode must be '%s' or '%s', got '%s'",
        M.FLAT,
        M.VOXEL,
        tostring(merged.mode)
      )
    )
  end

  merged.fov_y_degrees = clamp(
    require_number("fov_y_degrees", merged.fov_y_degrees),
    M.MIN_FOV_Y_DEGREES,
    M.MAX_FOV_Y_DEGREES
  )
  merged.depth_per_unit =
    clamp(require_number("depth_per_unit", merged.depth_per_unit), 0, M.MAX_DEPTH_PER_UNIT)
  merged.yaw_degrees = require_number("yaw_degrees", merged.yaw_degrees) % 360
  merged.voxel_max_width = math.floor(
    clamp(require_number("voxel_max_width", merged.voxel_max_width), 1, M.MAX_VOXEL_MAX_WIDTH)
  )
  merged.voxel_depth =
    math.floor(clamp(require_number("voxel_depth", merged.voxel_depth), 1, M.MAX_VOXEL_DEPTH))
  merged.light.ambient = clamp(require_number("light.ambient", merged.light.ambient), 0, 1)

  if #merged.light.direction ~= 3 then
    error("distract: render.light.direction must be three numbers")
  end
  for axis = 1, 3 do
    require_number("light.direction", merged.light.direction[axis])
  end

  return merged
end

--- The light direction as a unit vector, or straight down when the configured
--- direction has no length to normalise.
---@param settings table
---@return number[]
function M.light_direction(settings)
  local direction = settings.light and settings.light.direction or M.DEFAULTS.light.direction
  local length = math.sqrt(
    direction[1] * direction[1] + direction[2] * direction[2] + direction[3] * direction[3]
  )
  if length < 1e-9 then
    return { 0, 1, 0 }
  end
  return { direction[1] / length, direction[2] / length, direction[3] / length }
end

--- How one asset is drawn.
---
--- A manifest may pin itself to a mode — a speech bubble reads as flat overlay
--- furniture in a 3D session too — and everything else follows the configuration.
---@param settings table
---@param manifest table|nil
---@return string mode
function M.mode_for(settings, manifest)
  local declared = manifest and manifest.render
  if declared ~= nil then
    if not MODES[declared] then
      error(
        string.format(
          "distract: manifest '%s' declares render = '%s'; expected '%s' or '%s'",
          tostring(manifest.name),
          tostring(declared),
          M.FLAT,
          M.VOXEL
        )
      )
    end
    return declared
  end
  return settings.mode
end

---@param settings table
---@param manifest table|nil
---@return boolean
function M.is_voxel(settings, manifest)
  return M.mode_for(settings, manifest) == M.VOXEL
end

return M

File: engine/src/render.rs
//! What the renderer draws, and how.
//!
//! One settings block, pushed from Neovim exactly as the floor and the viewport
//! scope are, so the terminal backends and the overlay read the same numbers.
//! Nothing here measures anything for itself.

use serde::{Deserialize, Serialize};

use crate::camera::{Camera, DEFAULT_DEPTH_PER_UNIT, DEFAULT_FOV_Y_DEGREES};
use crate::voxel::{DEFAULT_DEPTH, DEFAULT_MAX_WIDTH, VoxelOptions};

/// How an entity is drawn.
#[derive(Debug, Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Default)]
#[serde(rename_all = "lowercase")]
pub enum RenderMode {
    /// One textured quad per entity, ordered by `z_index`. The default, and what
    /// every asset has always been drawn as.
    #[default]
    #[serde(rename = "2d", alias = "flat")]
    Flat,
    /// A voxel model extruded from the entity's current frame, depth-tested under
    /// a perspective camera.
    #[serde(rename = "3d", alias = "voxel")]
    Voxel,
}

impl RenderMode {
    pub fn is_voxel(self) -> bool {
        matches!(self, RenderMode::Voxel)
    }
}

/// Direction and strength of the single directional light.
///
/// One light, because the geometry is what gives a voxel pet its form; the
/// analytical multi-point model in `sprite_gen` is how a *flat* sprite gets one,
/// and running both would put two contradictory shading models on one asset.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct Light {
    /// Direction the light travels, in world axes. Normalised on use.
    pub direction: [f32; 3],
    /// Floor brightness for a face the light does not reach, 0..1. A face in
    /// full shadow at 0 is pure black, which reads as a hole in the model.
    pub ambient: f32,
}

impl Default for Light {
    fn default() -> Self {
        Self {
            direction: [-0.4, 0.8, -0.45],
            ambient: 0.42,
        }
    }
}

/// Everything the renderer needs that is not the world itself.
#[derive(Debug, Clone, Copy, PartialEq, Serialize, Deserialize)]
pub struct RenderSettings {
    pub mode: RenderMode,
    pub fov_y_degrees: f32,
    /// Depth of one unit of `z`, as a fraction of the eye distance.
    pub depth_per_unit: f32,
    /// How far a model is turned off head-on, in degrees.
    ///
    /// Zero renders a voxel pet face-on, where it is indistinguishable from its
    /// own sprite: the depth is there and nothing reveals it. A small turn is
    /// what makes the model read as a model.
    pub yaw_degrees: f32,
    pub voxel_max_width: u32,
    pub voxel_depth: u32,
    pub light: Light,
}

impl Default for RenderSettings {
    fn default() -> Self {
        Self {
            mode: RenderMode::Flat,
            fov_y_degrees: DEFAULT_FOV_Y_DEGREES,
            depth_per_unit: DEFAULT_DEPTH_PER_UNIT,
            yaw_degrees: 22.0,
            voxel_max_width: DEFAULT_MAX_WIDTH,
            voxel_depth: DEFAULT_DEPTH,
            light: Light::default(),
        }
    }
}

/// Bounds every field arrives with, because they arrive over IPC from a user's
/// configuration and an unbounded one is either a divide by zero or a mesh that
/// exhausts memory.
pub const MIN_FOV_Y_DEGREES: f32 = 10.0;
pub const MAX_FOV_Y_DEGREES: f32 = 120.0;
pub const MAX_VOXEL_MAX_WIDTH: u32 = 128;
pub const MAX_VOXEL_DEPTH: u32 = 64;
pub const MAX_DEPTH_PER_UNIT: f32 = 0.5;

impl RenderSettings {
    /// Clamps every field into its documented range.
    ///
    /// Called on the way in rather than at the point of use: a value that is
    /// wrong is wrong once, and clamping per frame would hide which field it was.
    pub fn sanitised(self) -> Self {
        Self {
            mode: self.mode,
            fov_y_degrees: self
                .fov_y_degrees
                .clamp(MIN_FOV_Y_DEGREES, MAX_FOV_Y_DEGREES),
            depth_per_unit: self.depth_per_unit.clamp(0.0, MAX_DEPTH_PER_UNIT),
            yaw_degrees: self.yaw_degrees % 360.0,
            voxel_max_width: self.voxel_max_width.clamp(1, MAX_VOXEL_MAX_WIDTH),
            voxel_depth: self.voxel_depth.clamp(1, MAX_VOXEL_DEPTH),
            light: Light {
                direction: self.light.direction,
                ambient: self.light.ambient.clamp(0.0, 1.0),
            },
        }
    }

    pub fn voxel_options(&self) -> VoxelOptions {
        VoxelOptions {
            max_width: self.voxel_max_width,
            depth: self.voxel_depth,
        }
    }

    /// The camera these settings describe, for a viewport of this size.
    ///
    /// Flat mode gets an orthographic camera, which reproduces the 2D pixel
    /// mapping exactly, so the depth buffer means the same thing in both modes.
    pub fn camera(&self, width: f32, height: f32) -> Camera {
        let mut camera = if self.mode.is_voxel() {
            Camera::perspective(width, height, self.fov_y_degrees)
        } else {
            Camera::orthographic(width, height)
        };
        camera.depth_per_unit = self.depth_per_unit;
        camera
    }

    /// The light direction as a unit vector, or straight down when the
    /// configured direction has no length to normalise.
    pub fn light_direction(&self) -> [f32; 3] {
        let [x, y, z] = self.light.direction;
        let length = (x * x + y * y + z * z).sqrt();
        if length < f32::EPSILON {
            return [0.0, 1.0, 0.0];
        }
        [x / length, y / length, z / length]
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn flat_is_the_default_mode() {
        assert_eq!(RenderSettings::default().mode, RenderMode::Flat);
        assert!(!RenderMode::default().is_voxel());
    }

    #[test]
    fn the_wire_names_are_the_ones_a_user_writes() {
        let voxel: RenderMode = serde_json::from_str("\"3d\"").expect("3d parses");
        let flat: RenderMode = serde_json::from_str("\"2d\"").expect("2d parses");
        assert_eq!(voxel, RenderMode::Voxel);
        assert_eq!(flat, RenderMode::Flat);
        assert_eq!(serde_json::to_string(&voxel).expect("encodes"), "\"3d\"");
    }

    #[test]
    fn every_field_is_clamped_into_its_range() {
        let wild = RenderSettings {
            mode: RenderMode::Voxel,
            fov_y_degrees: 5_000.0,
            depth_per_unit: 40.0,
            yaw_degrees: 20.0,
            voxel_max_width: 100_000,
            voxel_depth: 0,
            light: Light {
                direction: [0.0, 1.0, 0.0],
                ambient: 12.0,
            },
        }
        .sanitised();

        assert_eq!(wild.fov_y_degrees, MAX_FOV_Y_DEGREES);
        assert_eq!(wild.depth_per_unit, MAX_DEPTH_PER_UNIT);
        assert_eq!(wild.voxel_max_width, MAX_VOXEL_MAX_WIDTH);
        assert_eq!(wild.voxel_depth, 1, "a zero-depth slab has no geometry");
        assert_eq!(wild.light.ambient, 1.0);
    }

    #[test]
    fn flat_mode_asks_for_the_projection_that_matches_the_sprite_pass() {
        let camera = RenderSettings::default().camera(1600.0, 900.0);
        assert!(!camera.is_perspective());
    }

    #[test]
    fn voxel_mode_asks_for_a_perspective_camera_at_the_configured_field_of_view() {
        let settings = RenderSettings {
            mode: RenderMode::Voxel,
            fov_y_degrees: 60.0,
            ..Default::default()
        };
        let camera = settings.camera(1600.0, 900.0);
        assert!(camera.is_perspective());
        assert!((camera.eye_distance() - 450.0 / (30.0f32.to_radians()).tan()).abs() < 1e-3);
    }

    #[test]
    fn a_light_with_no_direction_falls_back_to_straight_down() {
        let settings = RenderSettings {
            light: Light {
                direction: [0.0, 0.0, 0.0],
                ambient: 0.4,
            },
            ..Default::default()
        };
        assert_eq!(settings.light_direction(), [0.0, 1.0, 0.0]);
    }

    #[test]
    fn the_light_direction_is_normalised() {
        let settings = RenderSettings {
            light: Light {
                direction: [0.0, 8.0, 0.0],
                ambient: 0.4,
            },
            ..Default::default()
        };
        assert_eq!(settings.light_direction(), [0.0, 1.0, 0.0]);
    }
}
